jobbergate-agent/jobbergate_agent/jobbergate/submit.py

```diff
@@ -72,26 +94,40 @@ async def submit_job_script(
 
         submit_dir = pending_job_submission.execution_directory or SETTINGS.DEFAULT_SLURM_WORK_DIR
 
-        job_script = None
-
-        for metadata in pending_job_submission.job_script.files:
-            local_script_path = submit_dir / metadata.filename
-            local_script_path.parent.mkdir(parents=True, exist_ok=True)
+        """
+        The flag `DOWNLOAD_JOB_SCRIPTS` controls whether the agent will download the job script files
+        to the submit directory. It should be set to `False` when the agent doesn't have permission
+        to write to the submit directory.
+        """
 
-            response = await backend_client.get(metadata.path)
-            response.raise_for_status()
-            local_script_path.write_bytes(response.content)
+        job_script = None
+        supporting_files = []
 
-            if metadata.file_type == FileType.ENTRYPOINT:
-                job_script = response.content.decode("utf-8")
+        for file in pending_job_submission.job_script.files:
+            if file.file_type == FileType.ENTRYPOINT:  # should have only one entrypoint
+                job_script = await retrieve_job_script_file(file, submit_dir)
 
-            logger.debug(f"Copied job script file to {local_script_path}")
+            elif file.file_type == FileType.SUPPORT:
+                supporting_files.append(file)
 
         JobSubmissionError.require_condition(
             job_script,
             "Could not find an executable script in retrieved job script data.",
         )
 
+        if SETTINGS.DOWNLOAD_JOB_SCRIPTS:
+            # Download the supporting files to the execution dir
+            files_to_download = [retrieve_job_script_file(file, submit_dir) for file in supporting_files]
+            await asyncio.gather(*files_to_download)
```

Comment:
This is also nice, to speed up the submissions. I feel job_scripts could be in this gather as well, to also take advantage of asyncio. Some changes may be required in the design here, but I hope it will simplify a bit the code.

Reply:
The main problem is that the supporting files need to be written but we don't care about their decoded data, as for the entrypoint we need the decoded data to pass as the payload.
I changed the logic to encapsulate everything related to the job script entrypoint in its own function.

Reply:
That is a fair point. It is looking good after the encapsulation.